types/wlr_layer_shell.c

```diff
@@ -287,7 +287,13 @@ static void handle_wlr_surface_committed(struct wlr_surface *wlr_surface,
 		surface->added = true;
 		wlr_signal_emit_safe(&surface->shell->events.new_surface,
 				surface);
-		assert(surface->output);
+		if (!surface->output) {
+			wl_resource_post_error(surface->resource,
```

Comment:
Sorry for the late review, I didn't notice you updated the PR.

Protocol errors should only be used for protocol violations. Here the client cannot guess that the compositor won't find any suitable output. So I think the compositor should instead close immediately the layer surface (and this can become an assert like `surface->output != NULL || surface->closed`).

Reply:
Oh, you already close it, good. So I guess what needs to be done here is do not send an error and re-add an assert.